File: onvif-gui/src/admintab.cpp
/*******************************************************************************
* admintab.cpp
*
* Copyright (c) 2020 Stephen Rhodes
*
* This program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License along
* with this program; if not, write to the Free Software Foundation, Inc.,
* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*
*******************************************************************************/

#include "camerapanel.h"

#include <QGridLayout>
#include <QLabel>
#include <QMessageBox>
#include <QProcess>
#include <QThreadPool>
#include <iostream>

AdminTab::AdminTab(QWidget *parent)
{
    cameraPanel = parent;

    textCameraName = new QLineEdit();
    textCameraName->setMaximumWidth(240);
    textAdminPassword = new QLineEdit();
    textAdminPassword->setMaximumWidth(240);
    buttonReboot = new QPushButton(tr("Reboot Camera"), this);
    buttonReboot->setMaximumWidth(160);
    buttonReboot->setEnabled(false);
    buttonSyncTime = new QPushButton(tr("Sync Time"), this);
    buttonSyncTime->setMaximumWidth(100);
    buttonLaunchBrowser = new QPushButton(tr("Browser"), this);
    buttonLaunchBrowser->setMaximumWidth(100);
    buttonHardReset = new QPushButton(tr("Hard Reset"), this);
    buttonHardReset->setMaximumWidth(160);
    buttonHardReset->setEnabled(false);
    checkEnableReboot = new QCheckBox("Enable Reboot");
    checkEnableReset = new QCheckBox("Enable Reset");

    lblCameraName = new QLabel("Camera Name");
    lblAdminPassword = new QLabel("Set admin Password");

    QGridLayout *layout = new QGridLayout();
    layout->addWidget(lblCameraName,        0, 0, 1, 1);
    layout->addWidget(textCameraName,       0, 1, 1, 2);
    layout->addWidget(lblAdminPassword,     1, 0, 1, 1);
    layout->addWidget(textAdminPassword,    1, 1, 1, 2);
    layout->addWidget(buttonReboot,         2, 0, 1, 1);
    layout->addWidget(buttonSyncTime,       2, 1, 1, 1);
    layout->addWidget(buttonHardReset,      2, 2, 1, 1);
    layout->addWidget(checkEnableReboot,    3, 0, 1, 1);
    layout->addWidget(buttonLaunchBrowser,  3, 1, 1, 1);
    layout->addWidget(checkEnableReset,     3, 2, 1, 1);
    setLayout(layout);

    connect(buttonLaunchBrowser, SIGNAL(clicked()), this, SLOT(launchBrowserClicked()));
    connect(checkEnableReboot, SIGNAL(clicked()), this, SLOT(enableRebootChecked()));
    connect(checkEnableReset, SIGNAL(clicked()), this, SLOT(enableResetChecked()));
    connect(buttonReboot, SIGNAL(clicked()), this, SLOT(rebootClicked()));
    connect(buttonHardReset, SIGNAL(clicked()), this, SLOT(hardResetClicked()));
    connect(buttonSyncTime, SIGNAL(clicked()), this, SLOT(syncTimeClicked()));
    connect(textCameraName, SIGNAL(textChanged(const QString &)), this, SLOT(onTextChanged(const QString&)));
    connect(textAdminPassword, SIGNAL(textChanged(const QString &)), this, SLOT(onTextChanged(const QString&)));

    connect(this, SIGNAL(updateFinished()), this, SLOT(onUpdateFinished()));

}

void AdminTab::update()
{
    onvif::Data onvif_data = CP->devices[CP->currentDataRow];
    QString camera_name = textCameraName->text();
    strncpy(onvif_data->camera_name, camera_name.toLatin1(), camera_name.length());
    onvif_data->camera_name[camera_name.length()] = '\0';
    setUser(textAdminPassword->text().toLatin1().data(), onvif_data);
    CP->btnApply->setEnabled(false);
    CP->refreshList();
    CP->cameraNames->setValue(onvif_data->serial_number, onvif_data->camera_name);
    QString password = textAdminPassword->text();
    strncpy(onvif_data->password, password.toLatin1(), password.length());
}

void AdminTab::updated(const onvif::Data& onvif_data)
{
    CP->devices[CP->currentDataRow] = onvif_data;
    emit updateFinished();
    std::cout << "amdin update finished" << std::endl;
}

void AdminTab::onUpdateFinished()
{
    initialize();
    CP->cameraList->setEnabled(true);
    setActive(true);
}

void AdminTab::clear()
{
    textCameraName->setText("");
    textAdminPassword->setText("");
    checkEnableReboot->setChecked(false);
    checkEnableReset->setChecked(false);
}

void AdminTab::setActive(bool active)
{
    textCameraName->setEnabled(active);
    textAdminPassword->setEnabled(active);
    buttonLaunchBrowser->setEnabled(active);
    buttonSyncTime->setEnabled(active);
    checkEnableReboot->setEnabled(active);
    checkEnableReset->setEnabled(active);
    lblCameraName->setEnabled(active);
    lblAdminPassword->setEnabled(active);
}

bool AdminTab::hasBeenEdited()
{
    bool result = false;

    onvif::Data onvif_data = CP->devices[CP->currentDataRow];
    QString camera_name = onvif_data->camera_name;
    if (camera_name != textCameraName->text())
        result = true;

    if (textAdminPassword->text().length() > 0)
        result = true;

    return result;
}

void AdminTab::initialize()
{
    onvif::Data onvif_data = CP->devices[CP->currentDataRow];
    textCameraName->setText(tr(onvif_data->camera_name));
    buttonReboot->setEnabled(false);
    buttonHardReset->setEnabled(false);
    checkEnableReboot->setChecked(false);
    checkEnableReset->setChecked(false);
}

void AdminTab::launchBrowserClicked()
{
    onvif::Data onvif_data = CP->devices[CP->currentDataRow];
    char host[128];
    extractHost(onvif_data->xaddrs, host);
#ifdef _WIN32
    QString cmd("\"C:\\Program Files\\Internet Explorer\\iexplore.exe\"");
#else
    QString cmd("xdg-open");
#endif 
    QStringList args = { QString("http://") + host};
    process.start(cmd, args);
}

void AdminTab::enableRebootChecked()
{
    buttonReboot->setEnabled(checkEnableReboot->isChecked());
}

void AdminTab::enableResetChecked()
{
    buttonHardReset->setEnabled(checkEnableReset->isChecked());
}

void AdminTab::rebootClicked()
{
    QMessageBox::StandardButton result = QMessageBox::question(this, "onvif-gui", "You are about to reboot the camera\nAre you sure you want to do this");
    if (result == QMessageBox::Yes) {
        onvif::Manager onvifBoss;
        onvifBoss.startReboot(CP->devices[CP->currentDataRow],
                            [&](const onvif::Data& onvif_data) { updated(onvif_data); });

        CP->btnApply->setEnabled(false);
        CP->cameraList->setEnabled(false);
        setActive(false);
    }
}

void AdminTab::hardResetClicked()
{
    QMessageBox::StandardButton result = QMessageBox::question(this, "onvif-gui", "You are about to HARD RESET the camera\nAll settings will be returned to default factory configuration\nAre you sure you want to do this");
    if (result == QMessageBox::Yes) {
        onvif::Manager onvifBoss;
        onvifBoss.startReset(CP->devices[CP->currentDataRow],
                            [&](const onvif::Data& onvif_data) { updated(onvif_data); });

        CP->btnApply->setEnabled(false);
        CP->cameraList->setEnabled(false);
        setActive(false);
    }
}

void AdminTab::syncTimeClicked()
{
    onvif::Manager onvifBoss;
    onvifBoss.startUpdateTime(CP->devices[CP->currentDataRow],
                        [&](const onvif::Data& onvif_data) { updated(onvif_data); });

    CP->btnApply->setEnabled(false);
    CP->cameraList->setEnabled(false);
    setActive(false);
}

void AdminTab::doneRebooting()
{
    buttonReboot->setEnabled(false);
    checkEnableReboot->setChecked(false);
}

void AdminTab::doneResetting()
{
    buttonHardReset->setEnabled(false);
    checkEnableReset->setChecked(false);
}

void AdminTab::onTextChanged(const QString &)
{
    if (hasBeenEdited())
        CP->btnApply->setEnabled(true);
    else
        CP->btnApply->setEnabled(false);
}

File: python/admintab.py
import os
import sys
import numpy as np
import cv2
import platform
from time import sleep
from PyQt6.QtWidgets import QComboBox, QLineEdit, QSpinBox, \
QGridLayout, QWidget, QCheckBox, QLabel, QPushButton, QMessageBox
from PyQt6.QtCore import Qt, pyqtSignal, QObject, QProcess

sys.path.append("../build/libonvif")
import onvif

class AdminTab(QWidget):
    def __init__(self, cp):
        super().__init__()
        self.cp = cp
        self.process = QProcess()

        self.txtCameraName = QLineEdit()
        self.txtCameraName.textEdited.connect(self.cp.onEdit)
        lblCameraName = QLabel("Camera Name")
        self.txtAdminPassword = QLineEdit()
        self.txtAdminPassword.textEdited.connect(self.cp.onEdit)
        lblAdminPassword = QLabel("Admin Password")
        self.btnSyncTime = QPushButton("Sync Time")
        self.btnSyncTime.clicked.connect(self.btnSyncTimeClicked)
        self.btnReboot = QPushButton("Reboot")
        self.btnReboot.clicked.connect(self.btnRebootClicked)
        self.btnHardReset = QPushButton("Hard Reset")
        self.btnHardReset.clicked.connect(self.btnHardResetClicked)
        self.btnHardReset.setEnabled(False)
        self.btnBrowser = QPushButton("Browser")
        self.btnBrowser.clicked.connect(self.btnBrowserClicked)
        self.chkEnableReset = QCheckBox("Enable Reset")
        self.chkEnableReset.stateChanged.connect(self.chkEnableResetChanged)

        lytMain = QGridLayout(self)
        lytMain.addWidget(lblCameraName,         0, 0, 1, 1)
        lytMain.addWidget(self.txtCameraName,    0, 1, 1, 2)
        lytMain.addWidget(lblAdminPassword,      1, 0, 1, 1)
        lytMain.addWidget(self.txtAdminPassword, 1, 1, 1, 2)
        lytMain.addWidget(self.btnSyncTime,      2, 0, 1, 1)
        lytMain.addWidget(self.btnReboot,        2, 1, 1, 1)
        lytMain.addWidget(self.btnHardReset,     2, 2, 1, 1)
        lytMain.addWidget(self.btnBrowser,       3, 1, 1, 1)
        lytMain.addWidget(self.chkEnableReset,   3, 2, 1, 1)

    def fill(self, onvif_data):
        self.txtCameraName.setText(onvif_data.alias)
        self.setEnabled(True)

    def edited(self, onvif_data):
        result = False
        if self.isEnabled():
            if not onvif_data.alias == self.txtCameraName.text():
                result = True
            if len(self.txtAdminPassword.text()) > 0:
                result = True
        return result
    
    def update(self, onvif_data):
        if self.edited(onvif_data):
            print("admin tab update")
            if not onvif_data.alias == self.txtCameraName.text():
                onvif_data.alias = self.txtCameraName.text()
                self.cp.devices[self.cp.lstCamera.currentRow()] = onvif_data
                self.cp.lstCamera.currentItem().setText(onvif_data.alias)
                self.cp.settings.setValue(onvif_data.serial_number(), onvif_data.alias)
                print("editing camera alias")
            if len(self.txtAdminPassword.text()) > 0:
                result = QMessageBox.question(self, "Warning", "Please confirm camera password change")
                if result == QMessageBox.StandardButton.Yes:
                    self.cp.boss.onvif_data = onvif_data
                    self.cp.boss.new_password = self.txtAdminPassword.text()
                    self.cp.boss.startPySetUser()
                    self.txtAdminPassword.clear()
                    print("editing admin password")

    def btnRebootClicked(self):
        result = QMessageBox.question(self, "Warning", "Please confirm reboot")
        if result == QMessageBox.StandardButton.Yes:
            self.cp.boss.onvif_data = self.cp.devices[self.cp.lstCamera.currentRow()]
            self.cp.boss.startPyReboot()
        print("btnRebootClicked")

    def btnSyncTimeClicked(self):
        print("btnSyncTimeClicked")
        self.cp.boss.onvif_data = self.cp.devices[self.cp.lstCamera.currentRow()]
        self.cp.boss.startPyUpdateTime()

    def btnHardResetClicked(self):
        result = QMessageBox.question(self, "Warning", "** THIS WILL ERASE ALL SETTINGS **\nAre you sure you want to do this?")
        if result == QMessageBox.StandardButton.Yes:
            self.cp.boss.onvif_data = self.cp.devices[self.cp.lstCamera.currentRow()]
            self.cp.boss.startPyReset()
        print("btnHardResetClicked")

    def btnBrowserClicked(self):
        print("btnBrowserClicked")
        onvif_data = self.cp.devices[self.cp.lstCamera.currentRow()]
        if platform.system() == "Linux":
            cmd = "xdg-open"
        if platform.system() == "Windows":
            cmd = "\"C:\\Program Files\\Internet Explorer\\iexplore.exe\""
        args = "http://" + onvif_data.host()
        print(args)
        self.process.start(cmd, [args,])

    def chkEnableResetChanged(self, state):
        print("chkEnableResetChanged", state)
        self.btnHardReset.setEnabled(state)